rpc/rpc-wrappers.go

```diff
@@ -558,6 +572,32 @@ func NewValidationAuthorityServer(rpc Server, impl core.ValidationAuthority) (er
 		return
 	})
 
+	rpc.Handle(MethodPerformValidation, func(req []byte) (response []byte, err error) {
+		var vaReq performValidationRequest
+		if err = json.Unmarshal(req, &vaReq); err != nil {
+			// AUDIT[ Improper Messages ] 0786b6f2-91ca-4f48-9883-842a19084c64
+			improperMessage(MethodPerformValidation, err, req)
+			return nil, err
+		}
+
+		records, err := impl.PerformValidation(vaReq.Domain, vaReq.Challenge, vaReq.Authz)
+		// If the type of error was a ProblemDetails, we need to return
+		// both that and the records to the caller (so it can update
+		// the challenge / authz in the SA with the failing records).
+		// The least error-prone way of doing this is to send a struct
+		// as the RPC response and return a nil error on the RPC layer,
+		// then unpack that into (records, error) to the caller.
+		probs, ok := err.(*probs.ProblemDetails)
+		if !ok && err != nil {
+			return nil, err
+		}
+		jsonResp, err := json.Marshal(performValidationResponse{records, probs})
```

Comment:
Since `json.Marshal` won't mix return values this can just be `return json.Marshal(...)`